### ISSUE TITLE
Dartdoc should support arguments comments, metadata

### ISSUE THREAD

**Issue by matanlurey** [dart-lang/dartdoc#1259]
Using long sets of named arguments is more common in Dart, especially given it's a style guide preference for Flutter.

For example:

```dart
class AlcoholicState {
  /// Creates a new [AlcoholicState].
  ///
  /// Given a [alcoholic] and a number of [bars], represents different bars they can
  /// still go to (haven't been banned from).
  factory AlcoholicState({
    String alcoholic,
    int bars,
    List<String> allowed,
  }) = ...
}
```

It would be preferable to write:

```dart
class AlcoholicState {
  /// Creates a new [AlcoholicState] to represent an [alcoholic] wanting to go to [bars].
  factory FooState({
    /// Name of the alcoholic.
    @required
    String alcoholic,

    /// How many bars the alcoholic wants to attend tonight.
    @required
    int bars,

    /// Optional; where the alcoholic is still allowed entry to.
    List<String> allowed,
  }) = ...
}
```

When using an IDE or viewing in Dartdoc, you could have nice doc comments for these complex sets of arguments. One addendum is that if you use a field initializer (`this.`) we should automatically use the public API comment there:

```dart
class AlcoholicState {
  /// Name of the alcoholic.
  @required
  final String alcoholic;

  AlcoholicState({this.alcoholic});
}
```

**Comment by sethladd** [comment]
cc @Hixie

**Comment by sethladd** [comment]
@matanlurey did you have any ideas how this would look when generated into HTML?

**Comment by matanlurey** [comment]
I am not too creative - something similar to JavaDoc?

My main rationale here is actually for the IDE, when you type:

```dart
new SugarEaterState(...)
```

I'd like to see:

```
  |-----------------------------------|
  | Name of the sugar eater, required.|
  |-----------------------------------| 
  sugarEater:
```

**Comment by sethladd** [comment]
> My main rationale here is actually for the IDE

Ah, you might have to file a feature request with analyzer then?

**Comment by Hixie** [comment]
Oh, yeah, being able to annotate individual arguments would be great for some of our classes. (And yeah, this.foo should automatically default the arg to the docs for `foo`.)

**Comment by marcoms** [comment]
IMO this is a pretty significant area where dartdoc could improve. Since there is no way for an editor to parse argument documentation, not being able to get any argument-specific documentation when hovering over it in VSCode or any other editor degrades the development experience quite a lot compared to Typescript for example (which a lot of React Native and web devs are coming from, me included).

Instead you have to search through the whole function documentation, and for functions with lots of arguments this gets quite painful.

Interestingly VSCode manages to extract docs for constructor arguments that correspond to member variables which themselves have doc comments. It would be nice if there was a way to do this for methods or freestanding functions

@matanlurey's suggestion of allowing doc comments above the arguments themselves seems quite elegant. Alternatively dartdoc could just support the commonly used `@param` syntax.

**Comment by Jordan-Nelson** [comment]
Are there any updates in this issue? I believe this is making it much harder to library developers to maintain proper documentation.

For example, it is very hard to provide good documentation in factory constructors. One scenario in which I run into this frequently is when using `built_value`, which has become very popular. Below is an example from their examples dir.

```dart
abstract class VerySimpleValue
    implements Built<VerySimpleValue, VerySimpleValueBuilder> {
  static Serializer<VerySimpleValue> get serializer =>
      _$verySimpleValueSerializer;

  int get value;

  factory VerySimpleValue(int value) => _$VerySimpleValue._(value: value);
  VerySimpleValue._();
}
```

If `VerySimpleValue` is exposed as part of the public API of my library, I want to ensure it has proper docs. If I add the docs below, consumers of my library will not see that when they use the constructor.

```dart
abstract class VerySimpleValue
    implements Built<VerySimpleValue, VerySimpleValueBuilder> {
  static Serializer<VerySimpleValue> get serializer =>
      _$verySimpleValueSerializer;

  /// Sets the value for the object
  int get value;

  factory VerySimpleValue(int value) => _$VerySimpleValue._(value: value);
  VerySimpleValue._();
}
```

I could do something like below, but this is repetitive and not what consumers would expect since it doesn't match what the docs would look like if this was a non-factory constructor with `value` as a member of `VerySimpleValue`.

```dart
abstract class VerySimpleValue
    implements Built<VerySimpleValue, VerySimpleValueBuilder> {
  static Serializer<VerySimpleValue> get serializer =>
      _$verySimpleValueSerializer;

  /// Sets the value for the object
  int get value;

  /// Very simple value
  ///
  /// [value] sets the value for the object
  factory VerySimpleValue(int value) => _$VerySimpleValue._(value: value);
  VerySimpleValue._();
}
```

I could use a template/macro to get rid of the repetition, but the docs become somewhat clunky, and again - not what consumers expect.

What I really want is to do this:

  ```dart
abstract class VerySimpleValue
    implements Built<VerySimpleValue, VerySimpleValueBuilder> {
  static Serializer<VerySimpleValue> get serializer =>
      _$verySimpleValueSerializer;

  /// {@template very_simple_value.value}
  /// Sets the value for the object
  /// {@endtemplate}
  int get value;

  /// Very simple value ...
  ///
  /// [value] sets the value for ...
  factory VerySimpleValue({
    /// {@macro very_simple_value.value}
    required int value
  }) => _$VerySimpleValue._(value: value);
  VerySimpleValue._();
}
```

And then of course the doc for `value` would be visible to consumers using the factory constructor.

This would not involve any repeated code, and give consumers a good experience, with info about constructor arguments in a predictable location.

built_value and factory constructors are just one common use case I see, but this can be applied to many other cases. Take the simple class below.

```dart
class Foo {
  /// Foo class.
  /// 
  /// [name] does something ...
  const Foo({
    required this.value,
    required String name,
  }) : _name = 'foo_$name';

  /// doc for value
  final int value;

  final String _name;
}
```

I believe this is how the dart docs would suggest this class is commented. But this isn't a great DX. The consumer of `Foo` has to look in two different places to find the doc comments for the two arguments to the constructor.

**Comment by srawlins** [comment]
No updates to this issue. I agree it would be a great enhancement.

**Comment by Jordan-Nelson** [comment]
Okay. Thanks for the quick response.

**Comment by Jordan-Nelson** [comment]
I want to add a more common example of where the DX could be improved incase it helps bump the priority of this.

`ThemeData` is a class most flutter developers use. The commonly used constructors are factory constructors. The `ThemeData()` factory constructor accepts 83 arguments, not counting deprecated arguments. There are great API docs for these, but you have to search though the source code to find them.

**Comment by navaronbracke** [comment]
Has there been more work on this? With the new JS interop functionality, this is something I would like to see being implemented.

Especially since there is now a concrete use case, documenting external factory constructors for anonymous types.

Today you _can_ get the parameters documented if they have fields, but external factories might not have corresponding fields.

I.e.

```dart
import 'dart:js_interop';

/// Create a JS Object like `{foo: 'some sring', bar: 42}`,
/// using static interop.
@JS()
@anonymous
@staticInterop
class MyAnonType {
  external factory MyAnonType({
    String? foo,
    int bar,
  });
}
```

**Comment by srawlins** [comment]
There has not been more work on this.

**Comment by g-apparence** [comment]
This could help many plugins and packages.

**Comment by CHB61** [comment]
Interested to see this supported.

It would also be helpful for widgets like TextFormField where most parameters don't have dartdocs available since they are not class variables. Instead the dartdoc on the constructor says: **"For documentation about the various parameters, see the [TextField] class and [TextField.new], the constructor"**.

In my case, I would like to do something like this:

```dart
class SelectableList<T> extends StatefulWidget {
  /// The initial value for a list.
  final List<T>? initialValueList;

  /// The initial value for a single item.
  final T? initialValueSingle;

  const SelectableList.single({
    super.key,
    /// this doesn't work, but would be nice if I could point to the respective class variable's dartdoc
    T? initialValue,
  })  : initialValueList = null,
        initialValueSingle = initialValue;

  const SelectableList.multi({
    super.key,
    List<T>? initialValue,
  })  : initialValueList = initialValue,
        initialValueSingle = null;

  @override
  State<SelectableList<T>> createState() => _SelectableListState<T>();
}
```

**Comment by dcharkes** [comment]
I'd love for this to be addressed.

It seems to be a rather common pattern to have default values that cannot be expressed as const on `this.` parameters in constructors. Whether it's widgets, or something structured in a similar approach. (See a rather long-winded write-up here: https://github.com/dart-lang/native/issues/2555#issuecomment-3234901893 It seems an unwanted outcome to complicate the actual Dart code due to a lack of dartdoc on parameters.)

cc @szakarias for dartdoc @DanTup for vscode plugin.

**Comment by sigurdm** [comment]
@johnniwinther what would it take to have the analyzer parse dartdoc on arguments? I guess that would be the first step

**Comment by DanTup** [comment]
> @DanTup for vscode plugin.

For VS Code, everything would be done by the analyzer/analysis server. There's a similar issue at https://github.com/dart-lang/sdk/issues/39746 that was originally about dartdocs on nested functions (so that references in them are detected and updated if symbols are renamed), but also has some discussion about allowing them elsewhere.

**Comment by srawlins** [comment]
I hate to say it, but I think we'll want the language team's blessing on adding a spot where doc comments can be found. They may say, "go for it; we don't want to get particular about that." But doc comments are mentioned in [the spec](https://storage.googleapis.com/dart-specification/DartLangSpecDraft.pdf).

Otherwise, I think you're right @sigurdm , the first step is to parse these on parameters, and store them in analyzer's syntax tree.

**Comment by parlough** [comment]
I've long wanted this and personally support supporting parameter comments in the analysis server and generated docs. Many of Flutter's widget constructors are great case studies for how they could be helpful. I for some reason thought it was explicitly decided against, but can't remember where, perhaps due to the Effective Dart guidance of [DO use prose to explain parameters, return values, and exceptions](https://dart.dev/effective-dart/documentation#do-use-prose-to-explain-parameters-return-values-and-exceptions).

So with this change, we'd definitely need to consider how to update the guidance. For example, there'd need to be clear guidelines on how parameters should (and shouldn't) be discussed in the primary comment versus on each parameter.

\cc @munificent For your thoughts from an Effective Dart perspective.

**Comment by sigurdm** [comment]
@srawlins good point about needing to update the language itself! @eernstg @munificent @leafpetersen have you ever looked into adding dartdoc for parameters?

**Comment by leafpetersen** [comment]
I'm a little confused.  I just tried this out, and as best I can tell both front-ends allow comments (including doc comments) inside of parameter lists.  I'm surprised that the spec even mentions doc comments, is there any reason for this?  I'd be inclined to just remove that text. 

In any case, this seems like something we should support, especially given the plans around primary constructors.

**Comment by srawlins** [comment]
Yeah I get what you're saying, Leaf. Both front-ends "allow" comments, including doc comments, inside parameter lists. So we're all good on the front-ends accepting the syntax today.

I only ask because (a) I found that bit in the spec, haha, and (b) the analyzer only parses comments _as_ doc comments in certain declarations. I'm not sure if this was informed by some earlier spec or just... something the team decided. So today, in:

```dart
void f(
  /// comment refers to [f].
  int p
) {}
```

the analyzer accepts the comment as a plain vanilla comment, and does not feature it as a documentation comment on `int p`. The CFE never has anything interesting to do with end-of-line comments, whether they start with `///` or just `//` so it also happily accepts the text.

But all that aside, if the language spec does not need to have a stance on what "declarations" can have doc comments, or what the scope is in which they are resolved, we're all good.

Folks have additionally requested that we take a `///` comment _anywhere_ and parse it like a doc comment, linking references in square brackets etc, see https://github.com/dart-lang/sdk/issues/54307. This would include doc-like comments above local variable declarations, or even statements. This issue is related to such requests, and I think the Dart Model team is then free to define how and where doc comments or doc-like comments can be written.

**Comment by leafpetersen** [comment]
We talked about this in the language meeting this morning.  The consensus was that deciding where to interpret a comment as a doc comment, and how to interpret references within such a comment, should be left up to the tooling teams who are in the best position to understand the user needs.  We are of course happy to consult.

@eernstg is planning to remove the section of the spec which discusses where doc comments can appear and their scoping behavior, keeping only the specification of what constitutes a doc comment.

Does that sound good to everyone?

**Comment by johnniwinther** [comment]
cc @szakarias

**Comment by eernstg** [comment]
https://github.com/dart-lang/language/pull/4511.

**Comment by szakarias** [comment]
This feature request is becoming more relevant with the upcoming primary and declaring constructor feature. How would we otherwise document the fields? The question is then, should we :

a) only allow doc comments on parameters for these types of constructors, and then the doc would belong to the associated fields. 

b) support doc comments on parameters in general. In that case, as @parlough points out, we clash with the current guidance from Effective Dart, on using prose to explain parameters.

**Comment by srawlins** [comment]
I think even more. But we can start with a) and b) 😁 

This can be done incrementally, or not. But yes I think

a) requires no more input/design. We can just do it in the analyzer (once we start writing more support for primary constructors).

b) I think does require some guidance from Effective Dart committee (I don't know if that's a real thing, haha). Or at least a request for guidance; the response might be "do as you will; we can address Effective Dart later." For example, we can _allow_ doc comments on parameters, and maybe stop there. Or we can allow doc comments on parameters _and process them differently_ in dartdoc, like adding the text to the doc comment of the function declaration (method, constructor, top-level function). And we can do those without changing the Effective Dart _recommendation_ to use prose. 🤷 

c) another step, which I suppose do not concern dartdoc, but I'll write it here: As per https://github.com/dart-lang/sdk/issues/54307 we can also allow doc comments on _local_ function declarations, and local variable declarations. So this is still limited to _declarations_ for some definition of declarations.

d) literally allow any doc comment to be a doc comment. Above a statement, between two tokens in an expression, whatever. The goal here (and for c) is to enhance _editor support_, where we let people write references in square brackets, which link out, support renames, are guarded by the `comment_references` lint rule, etc.

I am in favor of all of a) through d), but again they can be done incrementally, and for b) it would be good first to reach out to Effective Dart, which is usually done by opening an issue against the site-www GitHub repo.

**Comment by munificent** [comment]
I guess I am the "Effective Dart" committee. :)

I'm strongly in favor of a) and then having dartdoc attach those comments to the corresponding field (and not show up as documentation for the parameter in the constructor's documentation).

I'm not sure what b) means by "support". Doc comments are comments. As far as the language is concerned, they can appear anyway. There's no way to stop that. :) Unless the proposal here is to add more UI to dartdoc to surface doc comments on parameters for arbitrary functions/constructors/whatever? If so, I personally don't think that would be super useful and wouldn't encourage it.

For c) and d), yes, I would generally be happy if square bracket references worked in any comment in a codebase and could be understood by analyzer for code navigation. I don't think they need to be surfaced in the dartdoc output in any way, though.

**Comment by sigurdm** [comment]
> I'm not sure what b) means by "support".

 "support" means that comments will be recognized by the analyzer and dartdoc - more specifically:

* `[]` references will be resolved in an appropriate scope.
* When you hover a reference to the arguments in the method body you get the documentation for that in the pop-up
* The comment will (somehow) be integrated into the dartdoc page for the containing function.
* While you write a method-call the IDE can show you documentation for the specific parameter you are currently filling in.

I **think** that is the kind of support this issue is discussing. Such that you instead of 

```dart
/// Calculates the square root.
///
/// [x] must be non-negative.
double sqrt(double x) {
  // ...
}
```

You could write

```dart
/// Calculates the square root.
double sqrt(
  /// Must be non-negative.
  double x) {
  // ...
}
```

@matanlurey @parlough @dcharkes can you confirm this style of dartdoc is what is being requested in this issue? Can you give examples of what you would love to write?

I personally like the style of describing the parameters in the main dartdoc of the method as we recommend currently. But I can see the attraction of dividing it into the parameters also. Especially when you have a loong list of parameters.

**Comment by dcharkes** [comment]
Instead of:

```dart
/// The generator that generates bindings for `dart:ffi` from C and Objective-C
/// headers.
final class FfiGenerator {
  /// The configuration for header parsing of [FfiGenerator].
  final Headers headers;

  /// Configuration for enums.
  final Enums enums;

  const FfiGenerator({
    this.headers = const Headers(),
    this.enums = Enums.excludeAll,
  });
```

Write

Instead of:

```dart
/// The generator that generates bindings for `dart:ffi` from C and Objective-C
/// headers.
final class FfiGenerator {
  /// The configuration for header parsing of [FfiGenerator].
  final Headers headers;

  /// Configuration for enums.
  late final Enums enums;

  FfiGenerator({
    this.headers = const Headers(),

    /// Configuration for enums.
    Enums? enums,
  }) {
     enums = /*...*/
  }
```

It is not always possible to construct a non-nullable const object for the default value. Some types do not support const (`Logger`, `Uri`, ...)

> "support" means that comments will be recognized by the analyzer and dartdoc - more specifically:
> 
> * `[]` references will be resolved in an appropriate scope.
> * When you hover a reference to the arguments in the method body you get the documentation for that in the pop-up
> * The comment will (somehow) be integrated into the dartdoc page for the containing function.
> * While you write a method-call the IDE can show you documentation for the specific parameter you are currently filling in.

1, 2, 4 yes!

And for 3, Yes! Possible "somehow":

The constructor or method should probably not show all the parameters:

https://api.flutter.dev/flutter/material/Scaffold-class.html#constructors

(Repeating the source of the parameters unformatted is completely unreadable.)

Only when you click through to the constructor or method, probably we want to list all parameters individually like we do for properties in a class:

https://api.flutter.dev/flutter/material/Scaffold/Scaffold.html

https://api.flutter.dev/flutter/material/Scaffold-class.html#instance-properties
